numalogic/connectors/druid/_druid.py

```diff
@@ -128,33 +158,130 @@ def fetch(
+        return df
 
     def raw_fetch(self, *args, **kwargs) -> pd.DataFrame:
         raise NotImplementedError
+
+    def chunked_fetch(
+        self,
+        datasource: str,
+        filter_keys: list[str],
+        filter_values: list[str],
+        dimensions: list[str],
+        delay: float = 3.0,
+        granularity: str = "minute",
+        aggregations: Optional[dict] = None,
+        post_aggregations: Optional[dict] = None,
+        group_by: Optional[list[str]] = None,
+        pivot: Optional[Pivot] = None,
+        hours: int = 24,
+        chunked_hours: int = 6,
+    ) -> pd.DataFrame:
+        """
+        Fetch data concurrently, and concatenate the results.
+
+        Args:
+        ------
+            datasource: Data source to query
+            filter_keys: keys
+            filter_values: values
+            dimensions: The dimensions to group by
+            delay: Added delay to the fetch query from current time.
+            granularity: Time bucket to aggregate data by hour, day, minute, etc.
+            aggregations: A map from aggregator name to one of the
+                ``pydruid.utils.aggregators`` e.g., ``doublesum``
+            post_aggregations: postaggregations map
+            group_by: List of columns to group by
+            pivot: Pivot configuration
+            hours: Hours from now to skip training.
+            chunked_hours: Hours to fetch in each chunk
+
+        Returns
+        -------
+            Fetched dataframe
+
+        Raises
+        ------
+            ValueError: If chunked_hours is less than 1
+        """
+        if chunked_hours < 1:
+            raise ValueError("chunked_hours should be integer and >= 1.")
+
+        _start_time = time.perf_counter()
+        filter_pairs = make_filter_pairs(filter_keys, filter_values)
+
+        hours_elapsed = 0
+        chunked_dfs = []
+        qparams = []
+
+        while hours_elapsed < hours:
+            ref_dt = datetime.now(pytz.utc) - timedelta(hours=hours_elapsed)
+            qparams.append(
+                build_params(
+                    datasource=datasource,
+                    dimensions=dimensions,
+                    filter_pairs=filter_pairs,
+                    granularity=granularity,
+                    hours=min(chunked_hours, hours - hours_elapsed),
+                    delay=delay,
+                    aggregations=aggregations,
+                    post_aggregations=post_aggregations,
+                    reference_dt=ref_dt,
+                )
+            )
+            hours_elapsed += chunked_hours
+
+        max_threads = min(_MAX_CONCURRENCY, len(qparams))
+        _LOGGER.debug("Fetching data concurrently with %s threads", max_threads)
+        with ThreadPoolExecutor(max_workers=max_threads) as executor:
+            futures = [executor.submit(self._fetch, **params) for params in qparams]
+            for future in as_completed(futures):
+                chunked_dfs.append(future.result())
```

Comment:
Can we guarantee the order here? I mean, since this is a timeseries data (we value the order of the chunked data) but we are making concurrent calls. And, after the concurrent calls, we are appending the data, is this guaranteed to be in the right order?

Do we also have to reverse the df here since we are querying from latest to old data[assuming we get data in that order]. Is there a possibility that we might trend in the timeseries if we train in this reverse order?

Reply:
I have a small question here (very relevant to what  @shashank10456 is asking ): 
I ran a quick experiment and I think as_completed() in above, we might not be getting back data in the same order of input.
With "as_completed" it will sort in order based on thread execution time. 

If I removed the "as_completed", I am getting back the data back in input order. This way, we can avoid sorting and other things to.

Reply:
[I see in the comments](https://github.com/numaproj/numalogic/pull/349#discussion_r1500482781)
You are already doing this. Please ignore this comment.

Reply:
No, but I still think that even without resetting index, we should ensure order. The latency difference will not be much at all.

Reply:
what do you mean by "latency difference will not be much"?